tools/ReadConsoleInputStream/ReadConsoleInputStream.cs

```diff
@@ -0,0 +1,194 @@
+            // to this stream simultaneously.
+            if (buffer.Length - offset < count)
+                throw new IndexOutOfRangeException("IndexOutOfRange_IORaceCondition");
+
+            int bytesRead;
+            int ret;
+
+            if (buffer.Length == 0)
+            {
+                bytesRead = 0;
+                ret = Win32Error.ERROR_SUCCESS;
+            }
+            else
+            {
+                var charsRead = 0;
+                bytesRead = 0;
+
+                var records = new Kernel32.INPUT_RECORD[BufferSize];
+
+                // begin input loop
+                waitForInput:
+
+                var readSuccess = Kernel32.ReadConsoleInput(_handle, records, 256, out var recordsRead);
+                Debug.WriteLine("Read {0} input record(s)", recordsRead);
+
+                if (readSuccess && recordsRead > 0)
+                {
+                    for (var index = 0; index < recordsRead; index++)
+                    {
+                        var record = records[index];
+
+                        if (record.EventType == Kernel32.EVENT_TYPE.KEY_EVENT)
+                        {
+                            // skip key up events - if not, every key will be duped in the stream
+                            if (record.Event.KeyEvent.bKeyDown == false) continue;
+
+                            // pack ucs-2/utf-16le/unicode chars into position in our byte[] buffer.
+                            var glyph = (ushort) record.Event.KeyEvent.uChar;
+
+                            var lsb = (byte) (glyph & 0xFFu);
+                            var msb = (byte) ((glyph >> 8) & 0xFFu);
+
+                            // ensure we accommodate key repeat counts
+                            for (var n = 0; n < record.Event.KeyEvent.wRepeatCount; n++)
+                            {
+                                buffer[offset + charsRead * BytesPerWChar] = lsb;
+                                buffer[offset + charsRead * BytesPerWChar + 1] = msb;
+
+                                charsRead++;
+                            }
+                        }
+                        else
+                        {
+                            // ignore focus events (not doing so makes debugging absolutely hilarious)
+                            if (record.EventType != Kernel32.EVENT_TYPE.FOCUS_EVENT)
+                            {
+                                // I assume success adding records - this is not so critical
+                                // if it is critical to you, loop on this with a miniscule delay
+                                _nonKeyEvents.TryAdd(record);
+                            }
+                        }
+                    }
+
+                    bytesRead = charsRead * BytesPerWChar;
+
+                    // we should continue to block if no chars read (KEY_EVENT)
+                    // even though non-key events were dispatched
+                    if (bytesRead == 0) goto waitForInput;
```

Comment:
> goto waitForInput; [](start = 39, length = 19)

oh boy. I know our code guidelines would never let us get away with a goto.

Reply:
Are you being sarcastic? It's the most efficient way to do this ;)